File: EvoPGA/codes/gbParse.go
//gbParse.go
//The program is used to retrieve all the annotated protein encoding genes (GeneID, GI_acc, Protein_ID), genome coordinate, strand, and protein sequences.

package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"regexp"
	"strings"
)

func main() {
	if len(os.Args) <= 1 {
		fmt.Printf("Please input the gb file that is going to be parsed!\n")
		return
	}
	gbFileName := os.Args[1]
	gbFile, openErr := os.Open(gbFileName)
	if openErr != nil {
		fmt.Printf("Error happened in opening %s!\n", gbFileName)
		return
	}
	defer gbFile.Close()

	inputReader := bufio.NewReader(gbFile)
	var seq, molType, geneAcc, gene, locusTag, product, geneID string
	var flag, flag0, flag2 int
	//	var gi,gene,prot,geneStart,geneEnd,strand string
	for {
		line, readErr := inputReader.ReadString('\n')
		pat01, _ := regexp.MatchString("  rRNA  ", line)
		pat02, _ := regexp.MatchString("  tRNA  ", line)
		pat03, _ := regexp.MatchString("  misc_feature  ", line)
		pat04, _ := regexp.MatchString("  ncRNA  ", line)
		patA, _ := regexp.MatchString("/gene=", line)
		patB, _ := regexp.MatchString("/locus_tag=", line)
		patC, _ := regexp.MatchString("/product=", line)
		patD, _ := regexp.MatchString("/EC_number=", line)
		patF, _ := regexp.MatchString("                     /pseudo", line)

		pat1, _ := regexp.MatchString("  CDS  ", line)
		patE, _ := regexp.MatchString("/translation=", line)
		//pat3, _ := regexp.MatchString("/db_xref=\"GeneID:",line)
		//pat4, _ := regexp.MatchString("/db_xref=\"GI:",line)
		//pat5, _ := regexp.MatchString("/protein_id=\"",line)
		if pat01 {
			molType = "rRNA"
			line = strings.Replace(line, "rRNA", "", 1)
			geneAcc = seqTrim(line)
			flag0 = 1
			gene = ""
			geneID = ""
			locusTag = ""
			product = ""
		} else if pat02 {
			molType = "tRNA"
			line = strings.Replace(line, "tRNA", "", 1)
			geneAcc = seqTrim(line)
			flag0 = 1
			gene = ""
			geneID = ""
			locusTag = ""
			product = ""
		} else if pat03 {
			molType = "misc_feature"
			line = strings.Replace(line, "misc_feature", "", 1)
			geneAcc = seqTrim(line)
			flag0 = 1
			gene = ""
			geneID = ""
			locusTag = ""
			product = ""
		} else if pat04 {
			molType = "ncRNA"
			line = strings.Replace(line, "ncRNA", "", 1)
			geneAcc = seqTrim(line)
			flag0 = 1
			gene = ""
			geneID = ""
			locusTag = ""
			product = ""
		} else if flag0 == 1 {
			if patA {
				gene = strings.Replace(line, "/gene=\"", "", -1)
				gene = strings.Replace(gene, "\"", "", -1)
				gene = seqTrim(gene)
			}
			if patB {
				locusTag = strings.Replace(line, "/locus_tag=\"", "", -1)
				locusTag = strings.Replace(locusTag, "\"", "", -1)
				locusTag = seqTrim(locusTag)
			}
			if patC {
				product = strings.Replace(line, "                     /product=\"", "", -1)
				product = strings.Replace(product, "\"", "", -1)
				product = strings.Replace(product, "\n", "", -1)
				flag0 = 0
				fmt.Printf("%s\t%s\t%s\t%s\t%s\t%s\t\n", geneAcc, molType, locusTag, geneID, gene, product)
				gene = ""
				geneID = ""
				locusTag = ""
				product = ""
			}
		}
		if pat1 {
			molType = "CDS"
			line = strings.Replace(line, "CDS", "", 1)
			line = seqTrim(line)
			geneAcc = seqTrim(line)
			flag0 = 0
			flag = 1
			gene = ""
			geneID = ""
			locusTag = ""
			product = ""
		} else if flag == 1 {
			if patA {
				gene = strings.Replace(line, "/gene=\"", "", -1)
				gene = strings.Replace(gene, "\"", "", -1)
				gene = seqTrim(gene)
			}
			if patB {
				locusTag = strings.Replace(line, "/locus_tag=\"", "", -1)
				locusTag = strings.Replace(locusTag, "\"", "", -1)
				locusTag = seqTrim(locusTag)
			}
			if patC {
				product = strings.Replace(line, "                     /product=\"", "", -1)
				product = strings.Replace(product, "\"", "", -1)
				product = strings.Replace(product, "\n", "", -1)
			}
			if patD {
				geneID = strings.Replace(line, "/EC_number=\"", "", -1)
				geneID = strings.Replace(geneID, "\"", "", -1)
				geneID = seqTrim(geneID)
			}
			if patE {
				line = strings.Replace(line, "/translation=\"", "", -1)
				line = seqTrim(line)
				flag = 0
				if strings.Contains(line, "\"") {
					seq = strings.Replace(line, "\"", "", 1)
					fmt.Printf("%s\t%s\t%s\t%s\t%s\t%s\t%s\n", geneAcc, molType, locusTag, geneID, gene, product, seq)
					gene = ""
					geneID = ""
					locusTag = ""
					product = ""
					seq = ""
				} else {
					flag2 = 1
					seq = line
				}
			}
			if patF {
				fmt.Printf("%s\t%s\t%s\t%s\t%s\tpseudo\n", geneAcc, molType, locusTag, geneID, gene)
				gene = ""
				geneID = ""
				locusTag = ""
				product = ""
				seq = ""
				flag = 0
			}
		} else if flag2 == 1 {
			line = seqTrim(line)
			if strings.Contains(line, "\"") {
				line = strings.Replace(line, "\"", "", -1)
				seq += line
				fmt.Printf("%s\t%s\t%s\t%s\t%s\t%s\t%s\n", geneAcc, molType, locusTag, geneID, gene, product, seq)
				flag2 = 0
				gene = ""
				geneID = ""
				locusTag = ""
				product = ""
				seq = ""
			} else {
				seq += line
			}
		}
		if readErr == io.EOF {
			break
		}
	}
}

func seqTrim(s string) string {
	s = strings.Trim(s, "\r\n")
	s = strings.Trim(s, "\n")
	s = strings.Replace(s, " ", "", -1)
	return s
}

/* func substrPrint (s string, l int) string {
   var s0 string
   n := (len(s) - len(s)%l)/l
   for i:=0;i<n;i++ {
   	  s0 += s[i*l:(i+1)*l] + "\n"
   }

   if len(s)%l != 0 {
   	  s0 += s[l*n:len(s)] + "\n"
   }
   return s0
} */

File: EvoPGA/codes/gbParse.py
#!/usr/bin/env python3
"""
gbParse.py

Extract annotated protein-coding genes (CDS), rRNAs, tRNAs, ncRNAs, and misc_feature records from a GenBank file, including:

- genomic coordinates (geneAcc)
- feature type (molType)
- locus_tag
- GeneID (taken from /EC_number qualifier)
- gene name
- product description
- protein sequence (for CDS only)

Pseudo-genes (marked with /pseudo) are also reported.

Usage:
    python gbParse.py <genbank.gb> [rootDir]
"""

import os
import re
import sys

def seq_trim(s: str) -> str:
    return s.replace(" ", "").replace("\n", "").replace("\r", "")

def parse_genbank(gb_path: str):
    mol_type = ""
    gene_acc = ""
    gene = ""
    locus_tag = ""
    product = ""
    gene_id = ""
    seq = ""
    flag = 0  
    flag0 = 0  
    flag2 = 0  

    with open(gb_path, 'r') as f:
        for line in f:
            line = line.rstrip('\r\n')

            if re.search(r'\s+CDS\s+', line):
                mol_type = "CDS"
                gene_acc = seq_trim(line.replace("CDS", ""))
                flag = 1
                flag0 = 0
                gene = ""
                locus_tag = ""
                product = ""
                gene_id = ""
                seq = ""
            elif re.search(r'\s+rRNA\s+', line):
                mol_type = "rRNA"
                gene_acc = seq_trim(line.replace("rRNA", ""))
                flag0 = 1
                flag = 0
                gene = ""
                locus_tag = ""
                product = ""
                gene_id = ""
            elif re.search(r'\s+tRNA\s+', line):
                mol_type = "tRNA"
                gene_acc = seq_trim(line.replace("tRNA", ""))
                flag0 = 1
                flag = 0
                gene = ""
                locus_tag = ""
                product = ""
                gene_id = ""
            elif re.search(r'\s+misc_feature\s+', line):
                mol_type = "misc_feature"
                gene_acc = seq_trim(line.replace("misc_feature", ""))
                flag0 = 1
                flag = 0
                gene = ""
                locus_tag = ""
                product = ""
                gene_id = ""
            elif re.search(r'\s+ncRNA\s+', line):
                mol_type = "ncRNA"
                gene_acc = seq_trim(line.replace("ncRNA", ""))
                flag0 = 1
                flag = 0
                gene = ""
                locus_tag = ""
                product = ""
                gene_id = ""

            if flag0 == 1:
                if line.startswith("                     /gene="):
                    gene = line.split('"')[1]
                elif line.startswith("                     /locus_tag="):
                    locus_tag = line.split('"')[1]
                elif line.startswith("                     /product="):
                    product = line.split('"')[1]
                    flag0 = 0
                    print(f"{gene_acc}\t{mol_type}\t{locus_tag}\t{gene_id}\t{gene}\t{product}")

            elif flag == 1:
                if line.startswith("                     /gene="):
                    gene = line.split('"')[1]
                elif line.startswith("                     /locus_tag="):
                    locus_tag = line.split('"')[1]
                elif line.startswith("                     /product="):
                    product = line.split('"')[1]
                elif line.startswith("                     /EC_number="):
                    gene_id = line.split('"')[1]
                elif line.startswith("                     /pseudo"):
                    print(f"{gene_acc}\t{mol_type}\t{locus_tag}\t{gene_id}\t{gene}\tpseudo")
                    flag = 0
                    gene = ""
                    locus_tag = ""
                    product = ""
                    gene_id = ""
                    seq = ""
                elif line.startswith("                     /translation="):
                    seq = line.split('"')[1]
                    if seq.endswith('"'):
                        seq = seq[:-1]
                        print(f"{gene_acc}\t{mol_type}\t{locus_tag}\t{gene_id}\t{gene}\t{product}\t{seq}")
                        flag = 0
                        seq = ""
                    else:
                        flag2 = 1
                elif flag2 == 1:
                    if line.endswith('"'):
                        seq += line[:-1]
                        print(f"{gene_acc}\t{mol_type}\t{locus_tag}\t{gene_id}\t{gene}\t{product}\t{seq}")
                        flag2 = 0
                        flag = 0
                        seq = ""
                    else:
                        seq += seq_trim(line)

def main():
    root = "/home/yaozikun/Public_Dir/Bact_PGA/BactPGA/input"
    if len(sys.argv) >= 3:
        root = sys.argv[2]

    gb_path = os.path.join(root, sys.argv[1])
    if not os.path.isfile(gb_path):
        print(f"Error: Cannot open {gb_path}")
        sys.exit(1)

    parse_genbank(gb_path)

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python gbParse.py <genbank.gb> [rootDir]")
        sys.exit(1)
    main()